src/main/JayZ.ts

```diff
@@ -25,69 +31,99 @@ export class JayZ {
+        dataKey,
+      })
+
+      const __jayz__metadata: EncryptedItemMetadata<T, K> = {
+        encryptedDataKey,
+        nonce,
+        scheme: this.encryptor.scheme,
+        encryptedFieldNames: fieldsToEncrypt,
+      }
+
+      return {
+        ...encryptedItem,
+        __jayz__metadata,
+      }
     })
 
-    const __jayz__metadata: EncryptedItemMetadata<T, K> = {
-      encryptedDataKey,
-      nonce,
-      scheme: this.encryptor.scheme,
-      encryptedFieldNames: fieldsToEncrypt
-    }
+    return Promise.all(itemPromises)
+  }
 
-    return {
-      ...encryptedItem,
-      __jayz__metadata
+  async decryptItems<T, K extends keyof T>(
+    itemsToDecrypt: EncryptedJayZItem<T, K>[]
+  ): Promise<T[]> {
+    if (itemsToDecrypt.length === 0) {
+      return []
     }
-  }
 
-  async decryptItem<T, K extends keyof T>(
-    encryptedJayZItem: EncryptedJayZItem<T, K>
-  ): Promise<T> {
-    const {
-      nonce,
-      encryptedDataKey,
-      encryptedFieldNames
-    } = encryptedJayZItem.__jayz__metadata
-
-    const encryptedItem = { ...encryptedJayZItem }
-    delete encryptedItem.__jayz__metadata
-
-    const dataKey = await this.keyProvider.decryptDataKey(encryptedDataKey)
-
-    const { decryptedItem } = await this.encryptor.decrypt<T, K>({
-      encryptedItem,
-      fieldsToDecrypt: encryptedFieldNames,
-      dataKey,
-      nonce
+    const itemPromises = itemsToDecrypt.map(async (item) => {
+      const {
+        nonce,
+        encryptedDataKey,
+        encryptedFieldNames,
+      } = item.__jayz__metadata
+
+      const encryptedItem = { ...item }
+      delete (encryptedItem as any).__jayz__metadata
+
+      const dataKey = await this.keyProvider.decryptDataKey(encryptedDataKey)
+
+      const { decryptedItem } = await this.encryptor.decrypt<T, K>({
+        encryptedItem,
+        fieldsToDecrypt: encryptedFieldNames,
+        dataKey,
+        nonce,
+      })
+
+      memzero(dataKey)
+      return decryptedItem
     })
 
-    memzero(dataKey)
-    return decryptedItem
+    return Promise.all(itemPromises)
   }
 
-  private async getDataKey(): Promise<DataKey> {
-    if (this.dataKey === undefined) {
-      this.timesDataKeyUsed = 1
-      this.dataKey = await this.keyProvider.generateDataKey()
-    } else if (this.timesDataKeyUsed >= this.maxUsesPerDataKey) {
-      memzero(this.dataKey.dataKey)
-      this.timesDataKeyUsed = 1
-      this.dataKey = await this.keyProvider.generateDataKey()
-    } else {
-      this.timesDataKeyUsed += 1
-    }
+  private zipItemsWithDataKey<T, K extends keyof T>(
+    items: EncryptItemRequest<T, K>[]
+  ): Promise<{ itemToEncrypt: EncryptItemRequest<T, K>; key: DataKey }[]> {
+    const itemsWithDataKeys = items.map((itemToEncrypt) => {
+      if (
+        this.currentDataKey !== undefined &&
+        this.currentDataKeyUsesRemaining > 0
+      ) {
+        this.currentDataKeyUsesRemaining -= 1
+        return { itemToEncrypt, key: this.currentDataKey }
```

Comment:
Perhaps a nice thing to add here is a boolean flag that indicates if this key should be purged after encryption, which would allow us to `memzero` it safely.